examples/nlp/machine_translation/translate_ddp.py

```diff
@@ -0,0 +1,164 @@
+            reverse_lang_direction=args.reverse_lang_direction,
+        )
+    else:
+        dataset = TarredOneSideTranslationDataset(
+            text_tar_filepaths=args.text2translate,
+            metadata_path=args.metadata_path,
+            tokenizer=model.encoder_tokenizer,
+            shuffle_n=100,
+            shard_strategy="scatter",
+            world_size=world_size,
+            global_rank=rank,
+        )
+    loader = DataLoader(dataset, batch_size=1)
+    result_dir = os.path.join(args.result_dir, f'rank{rank}')
+    os.makedirs(result_dir, exist_ok=True)
+    originals_file_name = os.path.join(result_dir, 'originals.txt')
+    translations_file_name = os.path.join(result_dir, 'translations.txt')
+    num_translated_sentences = 0
+
+    if args.target_lang != 'zh':
+        target_detokenizer = MosesDetokenizer(lang=args.target_lang)
+    elif args.target_lang == 'zh':
+        target_detokenizer = PanguJiebaDetokenizer()
+
+    if args.target_lang or args.source_lang == 'ja':
+        sp_detokenizer = SentencePieceDetokenizer()
+
+    if args.source_lang != 'zh':
+        source_detokenizer = MosesDetokenizer(lang=args.source_lang)
+    elif args.source_lang == 'zh':
+        source_detokenizer = PanguJiebaDetokenizer()
+
+    with open(originals_file_name, 'w') as of, open(translations_file_name, 'w') as tf:
+        for batch_idx, batch in enumerate(loader):
+            for i in range(len(batch)):
+                if batch[i].ndim == 3:
+                    batch[i] = batch[i].squeeze(dim=0)
+                batch[i] = batch[i].to(rank)
+            if args.twoside:
+                src_ids, src_mask, _, _, _ = batch
+            else:
+                src_ids, src_mask = batch
+            if batch_idx % 100 == 0:
+                logging.info(
+                    f"{batch_idx} batches ({num_translated_sentences} sentences) were translated by process with "
+                    f"rank {rank}"
+                )
+            num_translated_sentences += len(src_ids)
+            translations = ddp_model(src_ids, src_mask, None, None)
+            translations = translations.cpu().numpy()
+            for t in translations:
+                t = model.decoder_tokenizer.ids_to_text(t)
+                if args.target_lang == 'ja':
+                    t = sp_detokenizer.detokenize(t.split())
+                translation = target_detokenizer.detokenize(t.split())
+                tf.write(translation + '\n')
+            for o in src_ids:
+                o = model.encoder_tokenizer.ids_to_text(o)
+                if args.target_lang == 'ja':
+                    o = sp_detokenizer.detokenize(o.split())
+                translation = source_detokenizer.detokenize(o.split())
+                of.write(translation + '\n')
+    cleanup()
+
+
+def main() -> None:
+    world_size = torch.cuda.device_count()
+    args = get_args()
+    mp.spawn(translate, args=(world_size, args), nprocs=world_size, join=True)
```

Comment:
I'm curious why you didn't use PyTorch Lightning `Trainer.test` for this? The idea behind using PTL is minimize having to write this kind of boiler plate code.

Reply:
Also, I'm actually having trouble getting this to run on my system with mixed GPUs. One of my GPUs is dedicated to graphics, but this script keeps trying to use it.